src/order/jordan_hoelder.lean

```diff
@@ -0,0 +1,794 @@
+{ F := _, coe := composition_series.series }
+
+instance [inhabited X] : inhabited (composition_series X) :=
+⟨{ length := 0,
+   series := λ _, default X,
+   step' := λ x, x.elim0 }⟩
+
+variables {X}
+
+lemma step (s : composition_series X) : ∀ i : fin s.length,
+  is_maximal (s i.cast_succ) (s i.succ) := s.step'
+
+@[simp] lemma coe_fn_mk (length : ℕ) (series step) :
+  (@composition_series.mk X _ _ length series step : fin length.succ → X) = series := rfl
+
+theorem lt_succ (s : composition_series X) (i : fin s.length) :
+  s i.cast_succ < s i.succ :=
+lt_of_is_maximal (s.step _)
+
+protected theorem strict_mono (s : composition_series X) : strict_mono s :=
+fin.strict_mono_iff_lt_succ.2 (λ i h, s.lt_succ ⟨i, nat.lt_of_succ_lt_succ h⟩)
+
+protected theorem injective (s : composition_series X) : function.injective s :=
+s.strict_mono.injective
+
+@[simp] protected theorem inj (s : composition_series X) {i j : fin s.length.succ} :
+  s i = s j ↔ i = j :=
+s.injective.eq_iff
+
+instance : has_mem X (composition_series X) :=
+⟨λ x s, x ∈ set.range s⟩
+
+lemma mem_def {x : X} {s : composition_series X} : x ∈ s ↔ x ∈ set.range s := iff.rfl
+
+lemma total {s : composition_series X} {x y : X} (hx : x ∈ s) (hy : y ∈ s) : x ≤ y ∨ y ≤ x :=
+begin
+  rcases set.mem_range.1 hx with ⟨i, rfl⟩,
+  rcases set.mem_range.1 hy with ⟨j, rfl⟩,
+  rw [s.strict_mono.le_iff_le, s.strict_mono.le_iff_le],
+  exact le_total i j
+end
+
+/-- The ordered `list X` of elements of a `composition_series X`. -/
+def to_list (s : composition_series X) : list X := list.of_fn s
+
+/-- Two `composition_series` are equal if they are the same length and
+have the same `i`th element for every `i` -/
+lemma ext_fun {s₁ s₂ : composition_series X}
+  (hl : s₁.length = s₂.length)
+  (h : ∀ i, s₁ i = s₂ (fin.cast (congr_arg nat.succ hl) i)) :
+  s₁ = s₂ :=
+begin
+  cases s₁, cases s₂,
+  dsimp at *,
+  subst hl,
+  simpa [function.funext_iff] using h
+end
+
+@[simp] lemma length_to_list (s : composition_series X) : s.to_list.length = s.length + 1 :=
+by rw [to_list, list.length_of_fn]
+
+lemma to_list_ne_nil (s : composition_series X) : s.to_list ≠ [] :=
+by rw [← list.length_pos_iff_ne_nil, length_to_list]; exact nat.succ_pos _
+
+lemma to_list_injective : function.injective (@composition_series.to_list X _ _) :=
+λ s₁ s₂ (h : list.of_fn s₁ = list.of_fn s₂),
+have h₁ : s₁.length = s₂.length,
+  from nat.succ_injective
+    ((list.length_of_fn s₁).symm.trans $
+      (congr_arg list.length h).trans $
+      list.length_of_fn s₂),
+have h₂ : ∀ i : fin s₁.length.succ, (s₁ i) = s₂ (fin.cast (congr_arg nat.succ h₁) i),
+  begin
+    assume i,
+    rw [← list.nth_le_of_fn s₁ i, ← list.nth_le_of_fn s₂],
+    simp [h]
+  end,
+begin
+  cases s₁, cases s₂,
+  dsimp at *,
+  subst h₁,
+  simp only [heq_iff_eq, eq_self_iff_true, true_and],
+  simp only [fin.cast_refl] at h₂,
+  exact funext h₂
+end
+
+lemma chain'_to_list (s : composition_series X) :
```

Comment:
Aah, I see that you did. Was this not practical as definition?

Reply:
I'm not sure if @ChrisHughes24 tried this, but I fiddled around with this myself, and it seems doable, although it's not as clean as it might appear (e.g., you have to keep a list_ne_nil assumption).

Reply:
Ooh, that `list_ne_nil` assumption is unfortunate. Where does it come from?

Reply:
For example, if you allow the empty list to be a composition series, then "composition_series.top" will need a list_ne_nil assumption (or a `nonempty X` assumption, in which case it returns a junk value if the composition series is empty).